multiuser/permission/che-multiuser-permission-workspace/src/main/java/org/eclipse/che/multiuser/permission/workspace/server/spi/jpa/MultiuserJpaWorkspaceDao.java

```diff
@@ -306,12 +311,62 @@ public void unsubscribe() {
 
     @Override
     public void onCascadeEvent(BeforeAccountRemovedEvent event) throws Exception {
+      boolean nonStoppedExists = false;
       for (WorkspaceImpl workspace :
           iterate(
               (maxItems, skipCount) ->
                   workspaceManager.getByNamespace(
                       event.getAccount().getName(), false, maxItems, skipCount))) {
-        workspaceManager.removeWorkspace(workspace.getId());
+        if (STOPPED.equals(workspace.getStatus())) {
+          workspaceManager.removeWorkspace(workspace.getId());
+          continue;
+        }
+
+        nonStoppedExists = true;
+
+        if (STOPPING.equals(workspace.getStatus())) {
+          // it's not possible to forcibly stop workspace. Continue check other and fail after this
+          // to retry
+          continue;
+        }
+
+        // workspace is RUNNING or STARTING. It's needed to stop them and remove after
+        tryStopAndRemoveWithSA(workspace);
+      }
+
+      if (!nonStoppedExists) {
+        // all workspace are already removed
+        return;
+      }
+
+      // There is at least one workspace that was marked to be stop/removed after stop.
+      // It's needed to check if it's finished already
+      for (WorkspaceImpl workspace :
+          iterate(
+              (maxItems, skipCount) ->
+                  workspaceManager.getByNamespace(
+                      event.getAccount().getName(), false, maxItems, skipCount))) {
+        if (!STOPPED.equals(workspace.getStatus())) {
+          throw new ConflictException("User has workspace that is not stopped yet. Try again");
+        }
+      }
+      // Every workspace that was marked as to remove after stop is already removed
+    }
+
+    /**
+     * Going to reset EnvironmentContext, in this case Service Account(SA) credentials will be used
+     * for stopping workspace. If SA have required permissions workspace will be stopped otherwise
+     * exception thrown.
+     */
+    private void tryStopAndRemoveWithSA(WorkspaceImpl workspace)
+        throws ServerException, NotFoundException, ConflictException {
+      EnvironmentContext current = EnvironmentContext.getCurrent();
+      try {
+        EnvironmentContext.reset();
+        // Stop workspace and remove it after it was stopped
+        workspaceManager.stopWorkspace(workspace.getId(), of(REMOVE_WORKSPACE_AFTER_STOP, "true"));
```

Comment:
`of()` is very non-descriptive. Would IMHO read better if this was `Map.of(...)`.